lightyear/src/netconnection/steam/server.rs

```diff
@@ -0,0 +1,186 @@
+use crate::netconnection::ClientId;
+use crate::prelude::Io;
+use steamworks::{
+    networking_sockets::{InvalidHandle, ListenSocket, NetConnection},
+    networking_types::{ListenSocketEvent, NetConnectionEnd, NetworkingConfigEntry, SendFlags},
+    Client, ClientManager, FriendFlags, Friends, LobbyId, Manager, Matchmaking, ServerManager,
+    SteamError, SteamId,
+};
+use tracing::error;
+
+pub enum AccessPermission {
+    /// Everyone can connect
+    Public,
+    /// No one can connect
+    Private,
+    /// Only friends from the host can connect
+    FriendsOnly,
+    /// Only user from this list can connect
+    InList(HashSet<SteamId>),
+    /// Users that are in the lobby can connect
+    InLobby(LobbyId),
+}
+
+pub struct SteamServerConfig {
+    pub max_clients: usize,
+    pub access_permission: AccessPermission,
+}
+
+#[cfg_attr(feature = "bevy", derive(bevy_ecs::system::Resource))]
+pub struct SteamNetServer {
+    server: steamworks::Server,
+    listen_socket: ListenSocket<ServerManager>,
+    max_clients: usize,
+    access_permission: AccessPermission,
+    connections: HashMap<ClientId, NetConnection<ServerManager>>,
+    packet_queue: VecDeque<(ReadWordBuffer, ClientId)>,
+}
+
+impl SteamNetServer {
+    pub fn new(
+        server: steamworks::Server,
+        config: SteamServerConfig,
+    ) -> Result<Self, InvalidHandle> {
+        Ok(Self {
+            server,
+            listen_socket,
+            max_clients: config.max_clients,
+            access_permission: config.access_permission,
+            connections: HashMap::new(),
+            packet_queue: Default::default(),
+        })
+    }
+
+    fn num_connected_clients(&self) -> usize {
+        self.connections.len()
+    }
+
+    fn max_clients(&self) -> usize {
+        self.max_clients
+    }
+
+    /// Update the access permission to the server,
+    /// this change only applies to new connections.
+    fn set_access_permissions(&mut self, access_permission: AccessPermission) {
+        self.access_permission = access_permission;
+    }
+
+    /// Disconnects a client from the server.
+    fn disconnect_client(&mut self, client_id: ClientId, flush_last_packets: bool) {
+        if let Some((_key, value)) = self.connections.remove_entry(&client_id) {
+            let _ = value.close(
+                NetConnectionEnd::AppGeneric,
+                Some("Client was kicked"),
+                flush_last_packets,
+            );
+        }
+    }
+
+    /// Disconnects all active clients including the host client from the server.
+    fn disconnect_all(&mut self, flush_last_packets: bool) {
+        self.connections.keys().for_each(|client_id| {
+            self.disconnect_client(*client_id, flush_last_packets);
+        });
+    }
+}
+
+impl NetServer for SteamNetServer {
+    /// Update server connections, and receive packets from the network.
+    fn try_update(&mut self, delta_ms: f64, io: &mut Io) -> Result<ConnectionEvents> {
```

Comment:
food for thought, maybe a later update

In order to avoid repeated unnecessary allocations this could return an iterator or reuse the `Vec`s in `ConnectionEvents`